### PR TITLE
Add MobileViTv2

### PR DESCRIPTION
**Pull request by shehanmunasinghe**
# What does this PR do?

Adds the MobileViTv2 model into transformers library (PS: Work in Progress)

Fixes # ([issue](https://github.com/huggingface/transformers/issues/22570))

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/22570 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

File: docs/source/en/serialization.mdx
Comment (amyeroberts) [inline docs/source/en/serialization.mdx:102]: Typo?

```suggestion
- MobileViTV2
```

File: src/transformers/models/mobilevitv2/image_processing_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/image_processing_mobilevitv2.py:234]:
```suggestion
    ) -> BatchFeature:
```

File: tests/models/mobilevitv2/test_modeling_mobilevitv2.py
Comment (amyeroberts) [inline tests/models/mobilevitv2/test_modeling_mobilevitv2.py:188]: I think you'll need to add `pipeline_model_mapping` here e.g. like [here](https://github.com/huggingface/transformers/blob/d95045717e1a5bd8ce71223b5b8920e27687dee4/tests/models/mobilevit/test_modeling_mobilevit.py#L188)

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model! 

Looking at the modeling file -all of the layers are copied from the original MobileVit. Is this correct? If this is the case, then we can do something similar to models like [Matcha](https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/matcha.mdx), where we add the model in the README and checkpoints are loaded in using the Pix2Struct architecture, but no new modeling file is added.

**Comment by shehanmunasinghe** [comment]
@amyeroberts please note that this is still a work in progress. I'll let you know when it is ready for your review.

**Comment by shehanmunasinghe** [comment]
@amyeroberts , this is now ready for your review.

**Comment by shehanmunasinghe** [comment]
@amyeroberts , any updates?

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:81]: * `clip` would be a better name here. `bound_fn` sounds like the function has been bound to something
* Let's put `value` as the first value to keep with similar functions torch and numpy's clip 
* For typing - if both float and ints are accepted input types,[ just `float` can be used](https://peps.python.org/pep-0484/#the-numeric-tower)

```suggestion
def clip(value: float, min_val: float = float("-inf"), max_val: float = float("inf")) -> float:
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:286]: Why have this arg if it's always True?

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:285]: This should be take from the passed in the config 
```suggestion
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:255]: It's confusing to have a layer called `SelfAttention` which performs also cross attention - it should either be just `Attention` or two separate layers. Look to other models in the library to see how self/cross attention layers logic is normally handled

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:684]: For all of the `MobileViTv2Layer` layer here and all below: 
* `attn_unit_dim` shouldn't be calculated form hard coded values here e.g. 128 and divisor etc. These values should be part of the configuration e.g. `make_divisible(config.x * config.width_multiplier, divisor=config.y),`
* `stride` argument can be remove - it's 2 for everything
* `n_attn_blocks` should be in the config 
* `patch_size` is taken from the config - arg should be remove and config used instead 
* `ffn_multiplier` can be removed as an arg - its 2 everywhere (and should come from the config)

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:643]: You don't need to wrap all of the `make_divisible` call with `int(...)` - the function already handles that

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:66]: This should have a copied from statement

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:570]: Why specify `pytorch` in the name here - the whole model is in pytorch

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:573]: clearer var names

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:583]: Same here re method name

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:586]: same here re clearer var names

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:610]:
```suggestion
        patches, output_size = self.unfolding_pytorch(features)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model! 

Apologies for the delay in review. I've done a first pass on the files, generally in a good shape and very clean code 🤗  There's a bit of reworking needing in the attention layer so will need likely need more than one final review to make sure all in line with the library patterns. 

General comments: 
* Model name: MobileVitV2 to be consistent with other models + camelcasing
* Individual layers shouldn't have *args and **kwargs 
* __repr__ implementations aren't needed for model classes, particularly private ones
* Explicit var names in code and in comments e.g. N -> batch_size
* Typically don't use asserts in the main modeling code - raise exceptions instead or remove
* Feature extractors are deprecated for vision models
* Use of config values and matching with the model implementation - specifically the activations don't seem to match
* Layers should, as much as possible, not have kwargs for their init and just use the config to instantiate themselves. There's a few values in the modeling file taken form the config and then passed along to all the layers e.g.

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (shehanmunasinghe) [reply src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:1039]: Activation from the config file is for the base model. This line of code is for the segmentation head, which uses a different activation. Here I'm following the MobileViT(v1) implementation.

File: README_zh-hans.md
Comment (amyeroberts) [reply README_zh-hans.md:342]: This hasn't been resolved

File: src/transformers/models/auto/image_processing_auto.py
Comment (amyeroberts) [inline src/transformers/models/auto/image_processing_auto.py:80]:
```suggestion
```

File: src/transformers/models/mobilevitv2/__init__.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/__init__.py:39]:
```suggestion
```

File: src/transformers/models/mobilevitv2/__init__.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/__init__.py:68]:
```suggestion

```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:244]: nit: can be put on one line 

```suggestion
    def __init__(self, config: MobileViTV2Config, embed_dim: int) -> None:
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:237]:
```suggestion
        config (`MobileVitv2Config`):
             Model configuration object
        embed_dim (`int`):
            `input_channels` from an expected input of size :math:`(batch_size, input_channels, height, width)`
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:295]: - All of the `_forward_self_attn` logic should go here and `_forward_self_attn` removed (it's superfluous)
- Forward arguments shouldn't take arbirary *args and **kwargs
- Rename `x` to `hidden_states` to be consistent with other `forward` methods

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:349]: To remove

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:354]: To remove

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:381]: To remove

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:376]: nit: can be put on one line

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:410]: Perhaps we should rename this class to `MobileViTV2Block` - I'll let you decide what you think is best

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:420]: Can these ever have the value `None` ?

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:483]: more explicit var names

```suggestion
        batch_size, in_channels, img_height, img_width = feature_map.shape
        patches = nn.functional.unfold(
            feature_map,
            kernel_size=(self.patch_height, self.patch_width),
            stride=(self.patch_height, self.patch_width),
        )
        patches = patches.reshape(batch_size, in_channels, self.patch_height * self.patch_width, -1)

        return patches, (img_height, img_width)
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:515]: more explicit var names here in place of single letters i.e. B -> batch_size

**Review by amyeroberts — COMMENTED** [review]
Looking good! 

Just a small few changes and then it's good to merge. Overall clean and easy to read - just make sure to remove all old comments and rework any logic where it makes sense e.g. removing `_forward_self_attn` now the method is redundant.

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py
Comment (shehanmunasinghe) [reply src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:410]: I think using `MobileViTV2Layer` is more consistent with the original MobileViT (v1) implementation.

Comment (amyeroberts) [reply src/transformers/models/mobilevitv2/modeling_mobilevitv2.py]: Sounds good!

**Comment by shehanmunasinghe** [comment]
@amyeroberts, thanks for your feedback and I have now applied the suggested changes.

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work adding this model!

**Pull request merged**
